File: IT/2/Multithreading/Esercizi/Java/Esercizio13.java
/* file Esercizio13.java */

import java.util.Random;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class Target implements Runnable {
	private Lock lock;
 
	public Target(Lock lock) {
		this.lock = lock;
	}

	@Override
	public void run() {
		lock.lock();
		try
		{
			Thread t = Thread.currentThread();
			String name = t.getName();

			System.out.println("Il thread " + name + " entra nel metodo run().");
			t.sleep(5000);
			System.out.println("Il thread " + name + " esce dal metodo run().");
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}
		finally {
			lock.unlock();
		}
	}
}

public class Esercizio13 {
	public static void main(String [] args) {
		Random randomizer = new Random(System.currentTimeMillis());
		// tra 2 e 10 thread
		int nThreads = randomizer.nextInt(9) + 2;

		System.out.println("Verranno creati " + nThreads + " thread.");

		final ReentrantLock lock = new ReentrantLock();

		for (int i = 0; i < nThreads; i++) {
			Thread t = new Thread(new Target(lock));
			t.start();
		}
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio13.py
# file esercizio13.py

from __future__ import print_function
from multiprocessing import Process, current_process, Lock
from random import randint
from time import sleep

def func(lock):
	with lock:
		name = current_process().name
		print("Il processo {} entra nella funzione target.".format(name))
		sleep(5)
		print("Il processo {} esce dalla funzione target.".format(name))

if __name__ == "__main__":
	lock = Lock()

	# tra 2 e 10 processi
	nProc = randint(2, 10)
	print("Verranno creati {} processi.".format(nProc))

	procs = [Process(target=func, args=(lock,)) for i in range(nProc)]

	for p in procs:
		p.start()

	for p in procs:
		p.join()

